wave_lang/kernel/wave/barriers.py

```diff
@@ -90,29 +105,45 @@ def add_shared_memory_barriers(
     if not graph:
         graph = trace.get_root_graph()
 
+    if info is None:
+        info = defaultdict(SharedMemoryBarrierInfo)
+
     for node in graph.nodes:
         custom = get_custom(node)
-        if is_shared_memory_op(custom):
-            if last_node is None:
-                last_node = custom
+        if mem := is_shared_memory_op(custom):
+            state = info[mem]
+            if isinstance(custom, GatherToLDS):
+                state.is_async = True
+
+            if state.last_node is None:
+                state.last_node = custom
+                continue
+
+            if not need_barrier(custom, state.last_node):
+                state.last_node = custom
                 continue
-            if need_barrier(custom, last_node) and not is_barrier_between(
-                last_node.fx_node, custom.fx_node
-            ):
-                is_async = isinstance(last_node, GatherToLDS)
+
+            if barrier := is_barrier_between(state.last_node.fx_node, custom.fx_node):
+                barrier = get_custom(barrier)
+                # Promote the barrier to wait for async ops
+                if state.is_async and not barrier.wait_async_ops:
+                    barrier.update_arg("wait_async_ops", True)
+            else:
                 # Synchronize after the write to shared memory before we read from it.
                 with graph.inserting_before(node):
-                    SharedMemoryBarrier(wait_async_ops=is_async).add_to_graph(graph)
-            last_node = custom
+                    SharedMemoryBarrier(wait_async_ops=state.is_async).add_to_graph(
+                        graph
+                    )
+
+            state.is_async = False
```

Comment:
what if your last_node/custom is GatherToLDS, is the final state.is_async still False?